### PR TITLE
Intrinsics States.JsonMerge

### PR DESCRIPTION
**Pull request by kbrock**
Part of #64

Ruby missing a `Boolean` is frustrating me more and more.

This adds `JsonMerge`.

Please share your thoughts on how to handle the Boolean Attribute.

The other solution I had was a multi argument that handles a crystal style argument and it went through each of them. But it still had more modification in the comparison portion of the check code

Another solution is to just specify Object, and then manually make a check, but this required duplication of the error message.

### CODE REVIEW

File: lib/floe/workflow/intrinsic_function/transformer.rb
Comment (Fryguy) [inline lib/floe/workflow/intrinsic_function/transformer.rb:12]: Super-mega-minor
```suggestion

        class Boolean
```

**Comment by Fryguy** [comment]
Wow...it's that simple!?  LOL I was so scared to do this one haha.

**Comment by Fryguy** [comment]
oops...accidental close.

File: lib/floe/workflow/intrinsic_function/transformer.rb
Comment (Fryguy) [inline lib/floe/workflow/intrinsic_function/transformer.rb:15]: I agree with Rubocop - not a fan of single-line methods, even if simple, so let's just blow this out.

File: lib/floe/workflow/intrinsic_function/transformer.rb
Comment (Fryguy) [inline lib/floe/workflow/intrinsic_function/transformer.rb:14]: Can you add more detail here?  What makes it backwards in your opinion?  Are there tests demostrating this backwardness?

File: lib/floe/workflow/intrinsic_function/transformer.rb
Comment (Fryguy) [inline lib/floe/workflow/intrinsic_function/transformer.rb:58]: I wonder if we can do this better with something like `===`, which can handle kind_of logic?

File: spec/workflow/intrinsic_function_spec.rb
Comment (Fryguy) [inline spec/workflow/intrinsic_function_spec.rb:466]: Please include failure tests for every parameter (i.e. non-Hash in positions 1 and 2)

File: spec/workflow/intrinsic_function_spec.rb
Comment (Fryguy) [inline spec/workflow/intrinsic_function_spec.rb:454]: This context is not needed, so it complicated the test - Can we remove it and just hardcode an empty context like we do for the other tests?

**Comment by Fryguy** [comment]
I think the addition of Boolean is great.  Only thought I had was maybe to implement `===`, then we can unify the check by switching from kind_of? to `===`.  Not sure it will work, but there's so many tests, that we'll know pretty quickly.  Even so, I'm happy to merge as is, and this can be a followup.

**Comment by Fryguy** [comment]
Overall LGTM :+1: - A few comment on specs, then this is good to go.

File: lib/floe/workflow/intrinsic_function/parser.rb
Comment (Fryguy) [inline lib/floe/workflow/intrinsic_function/parser.rb:67]: This order is not alphabetical - it currently matches the order as listed in the stateslanguage.net spec.  The same goes for the order of the methods within the transformer.

File: lib/floe/workflow/intrinsic_function/transformer.rb
Comment (kbrock) [reply lib/floe/workflow/intrinsic_function/transformer.rb:14]: we do `Boolean.kind_of?(obj)` instead of `obj.kind_of?(Boolean)`

That is why we have a weird.

I have renamed to `cover?` and it reads a little better

File: lib/floe/workflow/intrinsic_function/transformer.rb
Comment (kbrock) [reply lib/floe/workflow/intrinsic_function/transformer.rb:58]: ooh. we are not asking if type is a kind of Boolean, we are asking if type == Boolean.
Introducing `===` makes the tests fail

**Comment by kbrock** [comment]
Looks like we need to introduce `deep_merge` -- not sure if we want to bring in a gem or want to implement this. (I was surprised to see that `deep_merge` gem was one of yours. YAY)

The problem with `===` and `kind_of?` is the method is defined on `Object` not `Boolean`. And we will not get called in those instances.

@Fryguy what do you think about defining `Boolean`?
I'm thinking of adding Boolean to other places too.

**Comment by kbrock** [comment]
update:
- Rebase
- Introduce `Boolean` to remove special cases around `kind_of?`
- Implement `deep_merge`
- Add more specs.

update:
- Fixed spacing in hash (rubocop)

**Comment by Fryguy** [comment]
This need a rebase after #256

File: spec/workflow/intrinsic_function_spec.rb
Comment (Fryguy) [inline spec/workflow/intrinsic_function_spec.rb:454]:
```suggestion
      it "it merges with right hand precedence" do
```

File: spec/workflow/intrinsic_function_spec.rb
Comment (Fryguy) [inline spec/workflow/intrinsic_function_spec.rb:462]:
```suggestion
      it "it deep merges with right hand precedence" do
```

File: lib/floe/boolean.rb
Comment (Fryguy) [inline lib/floe/boolean.rb:2]: Not sure I like this here - Originally you had it under Floe::Workflow::IntrinsicFunction::Boolean, which made more sense to keep it contained there.  Also, not sure I like messing with the internals of TrueClass and FalseClass.

Would like another opinion though - @agrare ?

**Comment by kbrock** [comment]
Update:
- rebase (resolved conflict)
- fixed spelling
- added number of parameters check

Still unsure what to do about Boolean.
Thinking that a MultiType (ala Crystal `TrueClass|FalseClass` / `[TrueClass, FalseClass]`) may be the best route.

**Comment by Fryguy** [comment]
Coincidentally, I added "Union" types to my States.Format PR, because I needed that.  Perhaps you can use that after that one is merged?

**Comment by kbrock** [comment]
Update:
- dropped Boolean and used Union type

**Review by Fryguy — APPROVED** [review]

**Pull request merged**
